### PR TITLE
chore(headless): remove undefined properties from request after it is built rather than using spread syntax in builder function

### PR DESCRIPTION
**Pull request by fbeaudoincoveo**
https://coveord.atlassian.net/browse/KIT-3082

I only eliminated the spread syntax from the `buildProductRecommendationsRequest` for now to show what I have in mind.

If you agree with the approach I'm proposing, I'll apply it to every build request function and convert this draft to an actual PR.

### CODE REVIEW

**Comment by louis-bompart** [comment]
Can you demonstrate a false ✅ (in a comment or another commit if you wish).

I want to explore other options; the recursion makes my baguette sense tingle. (I fear stack overflows).

I'm leaning toward a divide-and-conquer approach with stricter typing on the divided function.

**Comment by fbeaudoincoveo** [comment]
> Can you demonstrate a false ✅ (in a comment or another commit if you wish).

@louis-bompart , do you mean a case where the spread syntax could cause typing issues? I'll add a comment, yes!

> I want to explore other options; the recursion makes my baguette sense tingle. (I fear stack overflows).

> I'm leaning toward a divide-and-conquer approach with stricter typing on the divided function.

Initially I wanted to use a non-recursive function and call it "à la carte" in the return statements of the various `build<RequestType>Request>` functions. E.g.,:

```typescript
export const buildProductRecommendationsRequest = async (
  s: StateNeededByGetProductRecommendations
): Promise<ProductRecommendationsRequest> => {
  return removeUndefinedProperties({
    accessToken: s.configuration.accessToken,
    organizationId: s.configuration.organizationId,
    url: s.configuration.search.apiBaseUrl,
    locale: s.configuration.search.locale,
    timezone: s.configuration.search.timezone,
    visitorId: s.configuration.analytics.enabled
      ? await getVisitorID(s.configuration.analytics)
      : undefined,
    recommendation: s.productRecommendations.id,
    numberOfResults: s.productRecommendations.maxNumberOfRecommendations,
    fieldsToInclude: [
      ...ProductRecommendationDefaultFields,
      ...(s.productRecommendations.additionalFields || []),
    ],
    mlParameters: removeUndefinedProperties({
      itemIds:
        s.productRecommendations.skus &&
        s.productRecommendations.skus.length > 0
          ? s.productRecommendations.skus
          : undefined,
      brandFilter: s.productRecommendations.filter.brand,
      categoryFilter: s.productRecommendations.filter.category,
    }),
    actionsHistory: s.configuration.analytics.enabled
      ? historyStore.getHistory()
      : [],
    context: s.context?.contextValues,
    dictionaryFieldContext: s.dictionaryFieldContext?.contextValues,
    searchHub: s.searchHub,
  };
});
```

This means we'd have control over how deep we want to go when removing undefined properties. E.g., in the above example I'm removing all top-level undefined properties, as well as the nested undefined properties in the `mlParameters` object, but I'm not removing nested undefined properties from elements in the `actionsHistory` array (which may be fine, as we're not doing this right now anyways).

File: packages/headless/src/features/search/search-request.ts [hunk 1]
```diff
@@ -70,6 +70,8 @@ export const buildSearchRequest = async (
       facetOptions: {freezeFacetOrder: state.facetOptions.freezeFacetOrder},
     }),
     ...(state.folding?.enabled && {
+      // TODO KIT-3082 remove the comment below; it's only to demonstrate how the spread syntax can cause issues
```
Comment (fbeaudoincoveo) [inline packages/headless/src/features/search/search-request.ts:73]: @louis-bompart if you uncomment this line and comment the next one, you'll see there are no compilation errors, and unit tests will all pass.

**Comment by fbeaudoincoveo** [comment]
Closing this for now; we won't have time to work on this for at least a sprint anwyays.

**Pull request closed without merging**